File: Modul-2-Encapsulation & Access Modifier/Tugas.py
class Buku:
    def __init__(self, judul, penulis, kode_buku, stok, lokasi_rak):
        self.judul = judul
        self.penulis = penulis
        self.kode_buku = kode_buku
        self._stok = stok
        self.__lokasi_rak = lokasi_rak
    def get_lokasi_rak(self):
        return self.__lokasi_rak
    def set_lokasi_rak(self, lokasi_baru):
        self.__lokasi_rak = lokasi_baru
    def tambah_stok(self, jumlah):
        self._stok += jumlah
    def kurangi_stok(self, jumlah):
        if self._stok >= jumlah:
            self._stok -= jumlah
        else:
            print("Stok tidak cukup!")
    def info_buku(self):
        return f"{self.judul} | Penulis: {self.penulis} | Kode: {self.kode_buku} | Stok: {self._stok} | Rak: {self.get_lokasi_rak()}"

class Peminjaman:
    def __init__(self, kode_buku, tanggal_pinjam, tanggal_kembali=None, status="Dipinjam"):
        self.kode_buku = kode_buku
        self.tanggal_pinjam = tanggal_pinjam
        self.tanggal_kembali = tanggal_kembali
        self.status = status

    def info_peminjaman(self):
        return f"Buku: {self.kode_buku} | Pinjam: {self.tanggal_pinjam} | Kembali: {self.tanggal_kembali} | Status: {self.status}"

class Anggota:
    def __init__(self, id_anggota, nama, maks_pinjam, status_aktif=True):
        self.id_anggota = id_anggota
        self.nama = nama
        self._maks_pinjam = maks_pinjam
        self.__status_aktif = status_aktif
        self.daftar_peminjaman = []  
    def get_status(self):
        return self.__status_aktif
    def set_status(self, status):
        self.__status_aktif = status
    def pinjam_buku(self, buku, tanggal_pinjam):
        if not self.__status_aktif:
            print(f"{self.nama} tidak aktif dan tidak dapat meminjam buku.")
            return
        if len(self.daftar_peminjaman) >= self._maks_pinjam:
            print(f"{self.nama} sudah mencapai batas peminjaman!")
            return
        if buku._stok <= 0:
            print(f"Stok buku '{buku.judul}' habis!")
            return
        buku.kurangi_stok(1)
        peminjaman = Peminjaman(buku.kode_buku, tanggal_pinjam)
        self.daftar_peminjaman.append(peminjaman)
    def kembalikan_buku(self, kode_buku, tanggal_kembali):
        for pem in self.daftar_peminjaman:
            if pem.kode_buku == kode_buku and pem.status == "Dipinjam":
                pem.status = "Dikembalikan"
                pem.tanggal_kembali = tanggal_kembali
                return
        print("Data peminjaman tidak ditemukan!")
    def info_anggota(self):
        return f"ID: {self.id_anggota} | Nama: {self.nama} | Status Aktif: {self.get_status()}"

b1 = Buku("Pemrograman Python", "Abror", "B001", 3, "Rak A1")
b2 = Buku("Basis Data", "Cakra", "B002", 2, "Rak B1")
b3 = Buku("Jaringan Komputer", "Ariel", "B003", 1, "Rak C1")

a1 = Anggota("A01", "Ariel", 3)
a2 = Anggota("A02", "Cakra", 2)

a1.pinjam_buku(b1, "2025-12-01")
a1.pinjam_buku(b2, "2025-12-01")

a2.pinjam_buku(b3, "2025-12-01")

a1.kembalikan_buku("B001", "2025-12-05")

delimiter = "=" * 50
print(delimiter)
print("Informasi Buku:")
print(b1.info_buku())
print(b2.info_buku())
print(b3.info_buku())

print(delimiter)
print("Informasi Anggota:")
print(a1.info_anggota())
print(a2.info_anggota())

print(delimiter)
print("Daftar Peminjaman Anggota 1:")
for p in a1.daftar_peminjaman:
    print(p.info_peminjaman())

print(delimiter)
print("Daftar Peminjaman Anggota 2:")
for p in a2.daftar_peminjaman:
    print(p.info_peminjaman())

File: Modul-2-Encapsulation & Access Modifier/tugas.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

class Buku {
public:
    string judul;
    string penulis;
    string kode_buku;
protected:
    int stok;              // protected
private:
    string lokasi_rak;     // private
public:
    Buku(string j, string p, string k, int s, string rak)
        : judul(j), penulis(p), kode_buku(k), stok(s), lokasi_rak(rak) {}
    // Getter lokasi rak
    string getLokasiRak() {
        return lokasi_rak;
    }
    // Setter lokasi rak
    void setLokasiRak(string rak) {
        lokasi_rak = rak;
    }
    // Tambah stok
    void tambahStok(int jumlah) {
        stok += jumlah;
    }
    // Kurangi stok
    bool kurangiStok(int jumlah) {
        if (stok >= jumlah) {
            stok -= jumlah;
            return true;
        }
        return false;
    }
    void infoBuku() {
        cout << judul << " | " << penulis 
             << " | Stok: " << stok 
             << " | Rak: " << lokasi_rak << endl;
    }
};

class Peminjaman {
public:
    string kode_buku;
    string tanggal_pinjam;
    string tanggal_kembali;
    string status;

    Peminjaman(string kode, string pinjam, string kembali = "-", string st = "Dipinjam")
        : kode_buku(kode), tanggal_pinjam(pinjam), tanggal_kembali(kembali), status(st) {}

    void infoPeminjaman() {
        cout << "Kode Buku: " << kode_buku
             << ", Pinjam: " << tanggal_pinjam
             << ", Kembali: " << tanggal_kembali
             << ", Status: " << status << endl;
    }
};

class Anggota {
public:
    string id_anggota;
    string nama;

protected:
    int maks_pinjam;        // protected

private:
    bool status_aktif;      // private

public:
    vector<Peminjaman> daftar_peminjaman;     // aggregation

    Anggota(string id, string n, int m, bool aktif = true)
        : id_anggota(id), nama(n), maks_pinjam(m), status_aktif(aktif) {}

    // Getter status
    bool getStatus() {
        return status_aktif;
    }

    // Setter status
    void setStatus(bool s) {
        status_aktif = s;
    }

    // Pinjam buku
    void pinjamBuku(Buku &buku, string tanggal) {
        if (!status_aktif) {
            cout << nama << " tidak aktif!" << endl;
            return;
        }

        if ((int)daftar_peminjaman.size() >= maks_pinjam) {
            cout << nama << " sudah mencapai batas maksimal peminjaman!" << endl;
            return;
        }

        if (buku.kurangiStok(1)) {
            daftar_peminjaman.push_back(Peminjaman(buku.kode_buku, tanggal));
            cout << nama << " berhasil meminjam " << buku.judul << endl;
        } else {
            cout << "Stok buku " << buku.judul << " habis!" << endl;
        }
    }

    // Kembalikan buku
    void kembalikanBuku(Buku &buku, string tanggal) {
        for (auto &p : daftar_peminjaman) {
            if (p.kode_buku == buku.kode_buku && p.status == "Dipinjam") {
                p.status = "Dikembalikan";
                p.tanggal_kembali = tanggal;
                buku.tambahStok(1);
                cout << nama << " mengembalikan " << buku.judul << endl;
                return;
            }
        }
        cout << "Buku tidak ditemukan dalam daftar peminjaman!" << endl;
    }

    void infoAnggota() {
        cout << "ID: " << id_anggota 
             << " | Nama: " << nama 
             << " | Status: " << (status_aktif ? "Aktif" : "Nonaktif") 
             << endl;
    }
};

class Perpustakaan {
public:
    vector<Buku*> daftar_buku;  // composition (mengelola objek Buku)

    void tambahBuku(Buku* b) {
        daftar_buku.push_back(b);
    }

    void infoBuku() {
        for (auto b : daftar_buku) {
            b->infoBuku();
        }
    }
};

int main() {

    Perpustakaan perpus;

    // 3 Buku
    Buku* b1 = new Buku("Python Dasar", "Abrar", "B001", 3, "Rak A1");
    Buku* b2 = new Buku("Pemrograman Dasar", "ariel", "B002", 2, "Rak A2");
    Buku* b3 = new Buku("Database SQL", "Cakra", "B003", 1, "Rak A3");

    perpus.tambahBuku(b1);
    perpus.tambahBuku(b2);
    perpus.tambahBuku(b3);

    // 2 Anggota
    Anggota a1("A01", "Abrar", 3);
    Anggota a2("A02", "Ariel", 2);

    // Peminjaman
    a1.pinjamBuku(*b1, "2025-01-01");
    a1.pinjamBuku(*b2, "2025-01-01");
    a2.pinjamBuku(*b3, "2025-01-02");

    // Pengembalian
    a1.kembalikanBuku(*b1, "2025-01-05");

    cout << "\n==============================\n";
    cout << "       INFORMASI BUKU\n";
    cout << "==============================\n";
    perpus.infoBuku();

    cout << "\n==============================\n";
    cout << "       INFORMASI ANGGOTA\n";
    cout << "==============================\n";
    a1.infoAnggota();
    a2.infoAnggota();

    cout << "\n==============================\n";
    cout << " DAFTAR PEMINJAMAN ANGGOTA 1\n";
    cout << "==============================\n";
    for (auto &p : a1.daftar_peminjaman) {
        p.infoPeminjaman();
    }

    cout << "\n==============================\n";
    cout << " DAFTAR PEMINJAMAN ANGGOTA 2\n";
    cout << "==============================\n";
    for (auto &p : a2.daftar_peminjaman) {
        p.infoPeminjaman();
    }

    return 0;
}
